File: frontend/src/sim/boards.ts
/**
 * Board registry. The simulator, canvas, wiring, and compiler were originally
 * hard-coded to the Arduino UNO (ATmega328P). This abstracts "which board" so
 * we can offer several. Phase A ships UNO + Nano (both ATmega328P, so the
 * avr8js configuration is identical and the in-browser simulation runs fully).
 * The Mega (ATmega2560) lands in phase B with its own hand-written avr8js
 * peripheral map.
 */

export type BoardId = 'uno' | 'nano' | 'mega'

export type Mcu = 'atmega328p' | 'atmega2560'

export interface BoardConfig {
  id: BoardId
  label: string
  /** DOM element id and the prefix the agent uses when wiring (e.g. UNO.D13). */
  canvasId: string
  /** Custom element from @wokwi/elements. */
  wokwiElement: string
  /** arduino-cli fully-qualified board name. */
  fqbn: string
  mcu: Mcu
  digitalPins: string[]
  analogPins: string[]
  pwmPins: string[]
  powerPins: string[]
}

const UNO_PWM = ['D3', 'D5', 'D6', 'D9', 'D10', 'D11']
const POWER = ['5V', '3V3', 'GND']

export const BOARDS: Record<BoardId, BoardConfig> = {
  uno: {
    id: 'uno',
    label: 'Arduino UNO',
    canvasId: 'UNO',
    wokwiElement: 'wokwi-arduino-uno',
    fqbn: 'arduino:avr:uno',
    mcu: 'atmega328p',
    digitalPins: Array.from({ length: 14 }, (_, i) => `D${i}`),
    analogPins: ['A0', 'A1', 'A2', 'A3', 'A4', 'A5'],
    pwmPins: UNO_PWM,
    powerPins: POWER,
  },
  nano: {
    id: 'nano',
    label: 'Arduino Nano',
    canvasId: 'NANO',
    wokwiElement: 'wokwi-arduino-nano',
    fqbn: 'arduino:avr:nano:cpu=atmega328',
    mcu: 'atmega328p',
    // Same ATmega328P as the UNO, but the Nano breaks out two extra
    // analog-only inputs (A6, A7).
    digitalPins: Array.from({ length: 14 }, (_, i) => `D${i}`),
    analogPins: ['A0', 'A1', 'A2', 'A3', 'A4', 'A5', 'A6', 'A7'],
    pwmPins: UNO_PWM,
    powerPins: POWER,
  },
  mega: {
    id: 'mega',
    label: 'Arduino Mega 2560',
    canvasId: 'MEGA',
    wokwiElement: 'wokwi-arduino-mega',
    fqbn: 'arduino:avr:mega:cpu=atmega2560',
    mcu: 'atmega2560',
    digitalPins: Array.from({ length: 54 }, (_, i) => `D${i}`),
    analogPins: Array.from({ length: 16 }, (_, i) => `A${i}`),
    pwmPins: [
      'D2', 'D3', 'D4', 'D5', 'D6', 'D7', 'D8', 'D9',
      'D10', 'D11', 'D12', 'D13', 'D44', 'D45', 'D46',
    ],
    powerPins: POWER,
  },
}

/** Boards a kid can pick today. Mega is registered above but its accurate
 * simulation arrives in phase B, so it is gated out of the picker for now. */
export const SELECTABLE_BOARDS: BoardId[] = ['uno', 'nano']

export const DEFAULT_BOARD: BoardId = 'uno'

export function getBoard(id: BoardId): BoardConfig {
  return BOARDS[id] ?? BOARDS[DEFAULT_BOARD]
}

/** All wireable pin names for a board (digital + analog + power). */
export function boardPins(board: BoardConfig): string[] {
  return [...board.digitalPins, ...board.analogPins, ...board.powerPins]
}

File: backend/app/boards.py
"""Board registry (backend mirror of frontend/src/sim/boards.ts).

The agent, pin validation, and the arduino-cli FQBN were originally hard-coded
to the Arduino UNO. This registry parameterizes "which board" so the same tool
surface serves the UNO, Nano (both ATmega328P), and the Mega (ATmega2560)."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Board:
    id: str
    label: str
    canvas_id: str  # DOM id + the prefix the agent wires to (e.g. "UNO")
    fqbn: str  # arduino-cli fully-qualified board name
    num_digital: int  # D0 .. D(num_digital-1)
    num_analog: int  # A0 .. A(num_analog-1)


BOARDS: dict[str, Board] = {
    "uno": Board("uno", "Arduino UNO", "UNO", "arduino:avr:uno", 14, 6),
    "nano": Board("nano", "Arduino Nano", "NANO", "arduino:avr:nano:cpu=atmega328", 14, 8),
    "mega": Board("mega", "Arduino Mega 2560", "MEGA", "arduino:avr:mega:cpu=atmega2560", 54, 16),
}

DEFAULT_BOARD = "uno"

_POWER_PINS = {
    "GND", "5V", "3V3", "3.3V", "VIN", "AREF", "IOREF", "RESET",
    "GND.1", "GND.2", "GND.3",
}


def get_board(board_id: str | None) -> Board:
    return BOARDS.get(board_id or DEFAULT_BOARD, BOARDS[DEFAULT_BOARD])


def board_ids() -> set[str]:
    return set(BOARDS)


def canvas_id_to_board(canvas_id: str) -> str | None:
    for b in BOARDS.values():
        if b.canvas_id == canvas_id:
            return b.id
    return None


def valid_board_pins(board_id: str) -> set[str]:
    """Acceptable pin tokens after the board id, e.g. {"D13", "13", "A0", "GND"}."""
    b = get_board(board_id)
    return (
        {f"D{n}" for n in range(b.num_digital)}
        | {str(n) for n in range(b.num_digital)}
        | {f"A{n}" for n in range(b.num_analog)}
        | set(_POWER_PINS)
    )
